### PR TITLE
Fix missing zero-offset sequences and add checking

### PR DESCRIPTION
**Pull request by SingleAccretion**
A couple changes:

1) Add a checker which ensures we do not forget to call `ExtendPtrVN` (or equivalent) when necessary.
2) Fix missing zero-offset sequences for `ADDR(LCL_VAR)` (no tests because reproducing bad codegen for this issue is "hard", and the checker asserts anyway).
3) Fix missing `NotAField` zero-offset sequences. These are important to preserve because they carry the fact we could have had an overlapping field in the sequence. Tests added.
4) Some printing enhancements for `NotAField` sequences, to make the dumps more informative.

Ref: #64773.

### CODE REVIEW

File: src/coreclr/jit/valuenum.h
Comment (SingleAccretion) [inline src/coreclr/jit/valuenum.h:187]: It is somewhat unfortunate, but MSVC unrolls loops better when using the iterator (as opposed to a static array that would been a simpler solution).

**Comment by SingleAccretion** [comment]
There is a typo in this change, but the proper version found two cases where we miss field sequences. So we'll leave this in the draft state until fixes land.

File: src/coreclr/jit/valuenum.h
Comment (jakobbotsch) [reply src/coreclr/jit/valuenum.h:187]: Does it really matter for a debug check function? I would rather keep the source simple.

Comment (SingleAccretion) [reply src/coreclr/jit/valuenum.h:187]: Not for debug checks, but we have other places where it would be nice to use it (and I am planning to): `fgValueNumberAddExceptionSetForOverflow`, `fgValueNumberAddExceptionSetForDivision`, `fgValueNumberAddExceptionSetForIndirection`.

Comment (jakobbotsch) [reply src/coreclr/jit/valuenum.h:187]: Ok. I know this is a draft, but it would be nice to land all the updates together with its introduction so that we don't end up doing it in multiple different ways.

Comment (SingleAccretion) [reply src/coreclr/jit/valuenum.h]: Left `VNKinds` out of this change.

**Comment by SingleAccretion** [comment]
This is now blocked on #64501.

**Comment by SingleAccretion** [comment]
Will assume OSX timeouts are not related (happening everywhere currently...).

[Diffs](https://dev.azure.com/dnceng/public/_build/results?buildId=1595639&view=ms.vss-build-web.run-extensions-tab) are cases where we were numbering a real-world array of [`StructWithOverlappingFields`](https://source.dot.net/#System.Runtime.Caching/System/Runtime/Caching/CacheUsage.cs,c414e5e1cd19af15,references)  (presumably, not in a way that lead to silent bad codegen).

@dotnet/jit-contrib

**Comment by jakobbotsch** [comment]
@SingleAccretion It looks like the fuzzlyn run hit an assert that was changed in #65387, can you investigate?

**Review by jakobbotsch — APPROVED** [review]
LGTM, thanks!

**Comment by SingleAccretion** [comment]
> can you investigate?

Yes, of course.

**Pull request merged**
